File: sections/100_grandmaster_finale/problems/07_virtual_tree_firewall/solution.py
import sys

INF = 10**40


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    index = 0
    n, query_count = data[index], data[index + 1]
    index += 2
    graph = [[] for _ in range(n)]
    for _ in range(n - 1):
        u, v, cost = data[index] - 1, data[index + 1] - 1, data[index + 2]
        index += 3
        graph[u].append((v, cost))
        graph[v].append((u, cost))

    log = max(1, n.bit_length())
    up = [[0] * n for _ in range(log)]
    minimum = [[INF] * n for _ in range(log)]
    depth = [0] * n
    tin = [0] * n
    tout = [0] * n
    timer = 0
    events = [(0, 0, INF, 0)]
    while events:
        u, parent, weight, phase = events.pop()
        if phase:
            tout[u] = timer
            continue
        tin[u] = timer
        timer += 1
        up[0][u] = parent
        minimum[0][u] = weight
        if u:
            depth[u] = depth[parent] + 1
        for bit in range(1, log):
            middle = up[bit - 1][u]
            up[bit][u] = up[bit - 1][middle]
            minimum[bit][u] = min(minimum[bit - 1][u], minimum[bit - 1][middle])
        events.append((u, parent, weight, 1))
        for v, cost in reversed(graph[u]):
            if v != parent:
                events.append((v, u, cost, 0))

    def is_ancestor(u, v):
        return tin[u] <= tin[v] and tout[v] <= tout[u]

    def lca(u, v):
        if is_ancestor(u, v):
            return u
        if is_ancestor(v, u):
            return v
        for bit in range(log - 1, -1, -1):
            if not is_ancestor(up[bit][u], v):
                u = up[bit][u]
        return up[0][u]

    def path_minimum(ancestor, vertex):
        difference = depth[vertex] - depth[ancestor]
        result = INF
        bit = 0
        while difference:
            if difference & 1:
                result = min(result, minimum[bit][vertex])
                vertex = up[bit][vertex]
            difference >>= 1
            bit += 1
        return result

    marked = [0] * n
    output = []
    for token in range(1, query_count + 1):
        k = data[index]
        index += 1
        terminals = [value - 1 for value in data[index:index + k]]
        index += k
        for vertex in terminals:
            marked[vertex] = token
        if k <= 1:
            output.append("0 0")
            continue
        terminals.sort(key=tin.__getitem__)
        nodes = terminals + [lca(terminals[i - 1], terminals[i]) for i in range(1, k)]
        nodes = sorted(set(nodes), key=tin.__getitem__)
        children = [[] for _ in nodes]
        stack = []
        for i, vertex in enumerate(nodes):
            while stack and not is_ancestor(nodes[stack[-1]], vertex):
                stack.pop()
            if stack:
                parent_index = stack[-1]
                children[parent_index].append(
                    (i, path_minimum(nodes[parent_index], vertex)))
            stack.append(i)

        dp_zero = [None] * len(nodes)
        dp_one = [None] * len(nodes)
        for i in range(len(nodes) - 1, -1, -1):
            if marked[nodes[i]] == token:
                zero, one = (INF, INF), (0, 0)
            else:
                zero, one = (0, 0), (INF, INF)
            for child, edge_cost in children[i]:
                cut_child = (dp_one[child][0] + edge_cost,
                             dp_one[child][1] + 1)
                separated = min(dp_zero[child], cut_child)
                new_zero = (zero[0] + separated[0], zero[1] + separated[1])
                new_one = min((one[0] + separated[0], one[1] + separated[1]),
                              (zero[0] + dp_one[child][0],
                               zero[1] + dp_one[child][1]))
                zero, one = new_zero, new_one
            dp_zero[i], dp_one[i] = zero, one
        answer = min(dp_zero[0], dp_one[0])
        output.append(f"{answer[0]} {answer[1]}")
    print("\n".join(output))


if __name__ == "__main__":
    main()

File: sections/100_grandmaster_finale/problems/07_virtual_tree_firewall/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;

constexpr int64 INF = 4'000'000'000'000'000'000LL;
using State = pair<int64, int>;

State add_state(State a, State b) {
    if (a.first >= INF || b.first >= INF) return {INF, INT_MAX / 2};
    return {a.first + b.first, a.second + b.second};
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, query_count;
    cin >> n >> query_count;
    vector<vector<pair<int, int64>>> graph(n);
    for (int i = 1; i < n; ++i) {
        int u, v;
        int64 cost;
        cin >> u >> v >> cost;
        --u;
        --v;
        graph[u].push_back({v, cost});
        graph[v].push_back({u, cost});
    }

    int log = max(1, static_cast<int>(bit_width(static_cast<unsigned>(n))));
    vector<vector<int>> up(log, vector<int>(n));
    vector<vector<int64>> minimum(log, vector<int64>(n, INF));
    vector<int> depth(n), tin(n), tout(n);
    struct Event { int u, parent; int64 weight; bool exit; };
    vector<Event> events{{0, 0, INF, false}};
    int timer = 0;
    while (!events.empty()) {
        Event event = events.back();
        events.pop_back();
        int u = event.u;
        if (event.exit) {
            tout[u] = timer;
            continue;
        }
        tin[u] = timer++;
        up[0][u] = event.parent;
        minimum[0][u] = event.weight;
        if (u != 0) depth[u] = depth[event.parent] + 1;
        for (int bit = 1; bit < log; ++bit) {
            up[bit][u] = up[bit - 1][up[bit - 1][u]];
            minimum[bit][u] = min(minimum[bit - 1][u],
                                  minimum[bit - 1][up[bit - 1][u]]);
        }
        events.push_back({u, event.parent, event.weight, true});
        for (auto it = graph[u].rbegin(); it != graph[u].rend(); ++it)
            if (it->first != event.parent)
                events.push_back({it->first, u, it->second, false});
    }

    auto is_ancestor = [&](int u, int v) {
        return tin[u] <= tin[v] && tout[v] <= tout[u];
    };
    auto lca = [&](int u, int v) {
        if (is_ancestor(u, v)) return u;
        if (is_ancestor(v, u)) return v;
        for (int bit = log - 1; bit >= 0; --bit)
            if (!is_ancestor(up[bit][u], v)) u = up[bit][u];
        return up[0][u];
    };
    auto path_minimum = [&](int ancestor, int vertex) {
        int difference = depth[vertex] - depth[ancestor];
        int64 result = INF;
        for (int bit = 0; bit < log; ++bit) {
            if (difference >> bit & 1) {
                result = min(result, minimum[bit][vertex]);
                vertex = up[bit][vertex];
            }
        }
        return result;
    };

    vector<int> marked(n);
    int token = 0;
    while (query_count--) {
        int k;
        cin >> k;
        vector<int> terminals(k);
        ++token;
        for (int &v : terminals) {
            cin >> v;
            --v;
            marked[v] = token;
        }
        if (k <= 1) {
            cout << "0 0\n";
            continue;
        }
        sort(terminals.begin(), terminals.end(),
             [&](int u, int v) { return tin[u] < tin[v]; });
        vector<int> nodes = terminals;
        for (int i = 1; i < k; ++i) nodes.push_back(lca(terminals[i - 1], terminals[i]));
        sort(nodes.begin(), nodes.end(), [&](int u, int v) { return tin[u] < tin[v]; });
        nodes.erase(unique(nodes.begin(), nodes.end()), nodes.end());

        vector<vector<pair<int, int64>>> virtual_children(nodes.size());
        vector<int> stack;
        for (int i = 0; i < static_cast<int>(nodes.size()); ++i) {
            while (!stack.empty() && !is_ancestor(nodes[stack.back()], nodes[i]))
                stack.pop_back();
            if (!stack.empty()) {
                int parent_index = stack.back();
                virtual_children[parent_index].push_back(
                    {i, path_minimum(nodes[parent_index], nodes[i])});
            }
            stack.push_back(i);
        }

        vector<State> dp_zero(nodes.size()), dp_one(nodes.size());
        for (int i = static_cast<int>(nodes.size()) - 1; i >= 0; --i) {
            State zero = marked[nodes[i]] == token ? State{INF, INT_MAX / 2}
                                                    : State{0, 0};
            State one = marked[nodes[i]] == token ? State{0, 0}
                                                   : State{INF, INT_MAX / 2};
            for (auto [child, edge_cost] : virtual_children[i]) {
                State separated = min(dp_zero[child],
                                      add_state(dp_one[child], {edge_cost, 1}));
                State new_zero = add_state(zero, separated);
                State new_one = min(add_state(one, separated),
                                    add_state(zero, dp_one[child]));
                zero = new_zero;
                one = new_one;
            }
            dp_zero[i] = zero;
            dp_one[i] = one;
        }
        State answer = min(dp_zero[0], dp_one[0]);
        cout << answer.first << ' ' << answer.second << '\n';
    }
}
